File: _static/code/lqcontrol/lq_permanent_1.jl
using QuantEcon
using Plots
using Plots.PlotMeasures
pyplot()

# == Model parameters == #
r = 0.05
β = 1/(1 + r)
T = 45
c_bar = 2.0
σ = 0.25
μ = 1.0
q = 1e6

# == Formulate as an LQ problem == #
Q = 1.0
R = zeros(2, 2)
Rf = zeros(2, 2); Rf[1, 1] = q
A = [1.0+r -c_bar+μ;
     0.0   1.0     ]
B = [-1.0; 0.0]
C = [σ; 0.0]

# == Compute solutions and simulate == #
lq = LQ(Q, R, A, B, C; bet=β, capT=T, rf=Rf)
x0 = [0.0; 1.0]
xp, up, wp = compute_sequence(lq, x0)

# == Convert back to assets, consumption and income == #
assets = vec(xp[1, :])               # a_t
c = vec(up + c_bar)                  # c_t
income = vec(σ * wp[1, 2:end] + μ)   # y_t

# == Plot results == #
p=plot(Vector[assets, c, zeros(T + 1), income, cumsum(income - μ)],
  lab=["assets" "consumption" "" "non-financial income" "cumulative unanticipated income"],
  color=[:blue :green :black :orange :red],
  xaxis=("Time"), layout=(2, 1),
  bottom_margin = 20mm, size=(600, 600))

File: _static/code/lqcontrol/lq_permanent_1.py
import numpy as np
import matplotlib.pyplot as plt
from quantecon import LQ

# == Model parameters == #
r = 0.05
β = 1/(1 + r)
T = 45
c_bar = 2
σ = 0.25
μ = 1
q = 1e6

# == Formulate as an LQ problem == #
Q = 1
R = np.zeros((2, 2))
Rf = np.zeros((2, 2))
Rf[0, 0] = q
A = [[1 + r, -c_bar + μ],
     [0,              1]]
B = [[-1],
     [ 0]]
C = [[σ],
     [0]]

# == Compute solutions and simulate == #
lq = LQ(Q, R, A, B, C, beta=β, T=T, Rf=Rf)
x0 = (0, 1)
xp, up, wp = lq.compute_sequence(x0)

# == Convert back to assets, consumption and income == #
assets = xp[0, :]           # a_t
c = up.flatten() + c_bar    # c_t
income = σ * wp[0, 1:] + μ  # y_t

# == Plot results == #
n_rows = 2
fig, axes = plt.subplots(n_rows, 1, figsize=(12, 10))

plt.subplots_adjust(hspace=0.5)

bbox = (0., 1.02, 1., .102)
legend_args = {'bbox_to_anchor': bbox, 'loc': 3, 'mode': 'expand'}
p_args = {'lw': 2, 'alpha': 0.7}

axes[0].plot(list(range(1, T+1)), income, 'g-', label="non-financial income",
             **p_args)
axes[0].plot(list(range(T)), c, 'k-', label="consumption", **p_args)

axes[1].plot(list(range(1, T+1)), np.cumsum(income - μ), 'r-',
             label="cumulative unanticipated income", **p_args)
axes[1].plot(list(range(T+1)), assets, 'b-', label="assets", **p_args)
axes[1].plot(list(range(T)), np.zeros(T), 'k-')

for ax in axes:
    ax.grid()
    ax.set_xlabel('Time')
    ax.legend(ncol=2, **legend_args)

plt.show()
